js/private/npm_import.bzl

```diff
@@ -0,0 +1,268 @@
+
+```starlark
+npm_import(
+    name = "npm__types_node-15.2.2",
+    package_name = "@types/node",
+    package_version = "15.12.2",
+    integrity = "sha512-zjQ69G564OCIWIOHSXyQEEDpdpGl+G348RAKY0XXy9Z5kU9Vzv1GMNnkar/ZJ8dzXB3COzD9Mo9NtRZ4xfgUww==",
+)
+```
+
+> This is similar to Bazel rules in other ecosystems named "_import" like
+> `apple_bundle_import`, `scala_import`, `java_import`, and `py_import`.
+> `go_repository` is also a model for this rule.
+
+The name of this repository should contain the version number, so that multiple versions of the same
+package don't collide.
+(Note that the npm ecosystem always supports multiple versions of a library depending on where
+it is required, unlike other languages like Go or Python.)
+
+To consume the downloaded package in rules, it must be "linked" into the link package in the
+package's `BUILD.bazel` file:
+
+```
+load("@npm__types_node-15.2.2//:nodejs_package.bzl", nodejs_package_types_node = "nodejs_package")
+
+nodejs_package_types_node()
+```
+
+The instantiates an `nodejs_binary` target for this package that can be referenced by the alias
+`@//link/package:npm__name` and `@//link/package:npm__@scope+name` for scoped packages.
+The `npm` prefix of these alias is configurable via the `namespace` attribute.
+
+When using `translate_pnpm_lock`, you can `link` all the npm dependencies in the lock files with:
+
+```
+load("@npm//:nodejs_packages.bzl", "nodejs_packages")
+
+nodejs_packages()
+```
+
+`translate_pnpm_lock` also creates convienence aliases in the external repository that reference
+the linked `nodejs_package` targets. For example, `@npm//name` and `@npm//@scope/name`.
+
+To change the proxy URL we use to fetch, configure the Bazel downloader:
+
+1. Make a file containing a rewrite rule like
+
+    rewrite (registry.nodejs.org)/(.*) artifactory.build.internal.net/artifactory/$1/$2
+
+1. To understand the rewrites, see [UrlRewriterConfig] in Bazel sources.
+
+1. Point bazel to the config with a line in .bazelrc like
+common --experimental_downloader_config=.bazel_downloader_config
+
+[UrlRewriterConfig]: https://github.com/bazelbuild/bazel/blob/4.2.1/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/UrlRewriterConfig.java#L66
+"""
+
+def is_windows_os(rctx):
+    os_name = rctx.os.name.lower()
+    return os_name.find("windows") != -1
+
+# TODO: handle npm packages with lifecycle hooks; src will be the output of the postinstall rule instead
+_NODEJS_PACKAGE_TMPL = \
+"""load("@aspect_rules_js//js:run_binary.bzl", "run_binary")
+
+def nodejs_package():
+    if "{link_package_guard}" != "." and native.package_name() != "{link_package_guard}":
+        fail("The nodejs_package() macro loaded from {nodejs_package_bzl} may only be called in the '{link_package_guard}' package. Move the call to the '{link_package_guard}' package BUILD file.")
+
+    run_binary(
+        name = "_{namespace}__{bazel_name}_postinstall",
```

Comment:
Lets make this
```
name = "{namespace}__{bazel_name}__postinstall",
```
so the target names are easier to reason about in a query
```
//example:example_npm_deps__at_gregmagolan_test-a_0.0.1
//example:example_npm_deps__at_gregmagolan_test-a_0.0.1__postinstall
//example:example_npm_deps__at_gregmagolan_test-b_0.0.2
//example:example_npm_deps__at_gregmagolan_test-b_0.0.2__dir
//example:example_npm_deps__at_gregmagolan_test-b_0.0.2__postinstall
//example:example_npm_deps__at_types_node_15.14.9
//example:example_npm_deps__at_types_node_15.14.9__dir
//example:example_npm_deps__at_types_node_15.14.9__postinstall
```
instead of the disjoint
```
//example:_example_npm_deps__at_gregmagolan_test-a_0.0.1_postinstall
//example:_example_npm_deps__at_gregmagolan_test-b_0.0.2_postinstall
//example:_example_npm_deps__at_types_node_15.14.9_postinstall
//example:example_npm_deps__at_gregmagolan_test-a_0.0.1
//example:example_npm_deps__at_gregmagolan_test-b_0.0.2
//example:example_npm_deps__at_gregmagolan_test-b_0.0.2__dir
//example:example_npm_deps__at_types_node_15.14.9
//example:example_npm_deps__at_types_node_15.14.9__dir
```